### PR TITLE
Refactor work loop 2

### PR DESCRIPTION
**Pull request by fw42**
WIP

second try at https://github.com/Shopify/resque/pull/16/
depends on https://github.com/resque/resque/pull/1470

### CODE REVIEW

**Comment by fw42** [comment]
@dylanahsmith, does this make sense to you?

File: lib/resque/worker.rb [hunk 1]
```diff
@@ -269,11 +234,31 @@ def work(interval = 5.0, &block)
...
+      procline "Processing #{job.queue} since #{Time.now.to_i} [#{job.payload_class_name}]"
+
+      log_with_severity :info, "got: #{job.inspect}"
+      job.worker = self
+
+      if fork_per_job?
+        perform_with_fork(job, &block)
       end
+
+      unless fork_per_job?
```
Comment (dylanahsmith) [inline lib/resque/worker.rb]: Now that the fork not implemented logic is moved into perform_with_fork, it isn't obvious in this method that `fork_per_job?` will change.  We should probably have perform_with_fork call `perform` when fork isn't implemented so  we can just use else for the `if fork_per_job?` condition here.

File: lib/resque/worker.rb [hunk 2]
```diff
@@ -303,6 +288,34 @@ def report_failed_job(job,exception)
...
+      run_hook :before_fork, job
+
+      begin
+        @child = fork do
+          unregister_signal_handlers if term_child
+          perform(job, &block)
+          exit! unless run_at_exit_hooks
+        end
+      rescue NotImplementedError
+        @fork_per_job = false
```
Comment (dylanahsmith) [inline lib/resque/worker.rb]: I would move the `return` into this rescue block.  This is also where I was suggesting that `perform` should get called.

Comment (fw42) [inline lib/resque/worker.rb]: done

File: lib/resque/worker.rb [hunk 3]
```diff
@@ -303,6 +286,36 @@ def report_failed_job(job,exception)
...
+        perform(job, &block)
+        return
+      end
+
+      return unless @child
```
Comment (dylanahsmith) [inline lib/resque/worker.rb]: I don't think this is needed now.

Comment (fw42) [inline lib/resque/worker.rb]: right, because the `fork` with block will exit after the block is run, and the child process will never get to this point?

Comment (dylanahsmith) [inline lib/resque/worker.rb]: right

**Comment by dylanahsmith** [comment]
:+1:

**Pull request closed without merging**